### ISSUE TITLE
Using separate functions to start stop streaming

### ISSUE THREAD

**Issue by rk-exxec** [alliedvision/VimbaPython#19]
Hi,
I'm using this library to display a video feed inside a Qt application. I have buttons to start and stop the video.

If I press the start button only one frame is acquired.
```python    
@Slot()
def start_preview(self):
    with self.vimba:
        with self.cam:
            self.cam.start_streaming(handler=self.frame_handler, buffer_count=10)
    self.is_running = True

@Slot()
def stop_preview(self):
    with self.vimba:
        with self.cam:
            self.cam.stop_streaming()
    self.is_running = False
```
I guess the streaming stops when I'm leaving the context manager?
Can I make the cam object persistent somehow? Or just keep the stream going?

**Comment by rk-exxec** [comment]
I used the enter and exit functions directly as a workaround.
```python
@Slot()
    def stop_preview(self):
        self.cam.stop_streaming()
        self.cam.__exit__(None, None, None)
        self.vimba.__exit__(None, None, None)
        self.is_running = False

    @Slot()
    def start_preview(self):
        self.vimba.__enter__()
        self.cam.__enter__()
        self.cam.start_streaming(handler=self.frame_handler, buffer_count=10)
        self.is_running = True
```
Would there be a more sophisticated approach to the problem? (I.e. an argument to the `start_streaming` function or something)

**Comment by NiklasKroeger-AlliedVision** [comment]
> I guess the streaming stops when I'm leaving the context manager?

That is correct. It happens because when you leave the context manager for the cam object, it calls [`__exit__`](https://github.com/alliedvision/VimbaPython/blob/master/vimba/camera.py#L370-L374) (as you already seem to know) which in turn calls [`_close`](https://github.com/alliedvision/VimbaPython/blob/master/vimba/camera.py#L935-L946) if no other context manager for the cam object is opened (`self.__context_cnt` is 0). `_close` then performs the actual call to `stop_streaming`.

> Can I make the cam object persistent somehow? Or just keep the stream going?
> I used the enter and exit functions directly as a workaround.

Using `__enter__` and `__exit__` in those functions definitely does what you want to achieve. But you do lose one important safety net: If your code throws an exception or is halted unexpectedly `__exit__` will most likely not be called and your camera connection will not be closed properly. This might lead to some unexpected behaviour when you try to connect to the camera again. So this approach is not ideal.

> Would there be a more sophisticated approach to the problem? (I.e. an argument to the start_streaming function or something)

What I would try to do is (in your `start_preview` method) spawn a separate thread which opens the `self.vimba` and `self.cam` context managers, starts the streaming and then waits for a [threading event](https://docs.python.org/3/library/threading.html#event-objects). When that event is set (by your `stop_preview` method) the thread would stop the streaming again and could leave the context managers.

This approach is demonstrated in the [multithreading_opencv.py](https://github.com/alliedvision/VimbaPython/blob/master/Examples/multithreading_opencv.py) example. Here you have a [`FrameProducer` class](https://github.com/alliedvision/VimbaPython/blob/master/Examples/multithreading_opencv.py#L126) which is created for every connected camera. It's job is to start the streaming of the camera and be the frame handler for the arriving frames. Once the camera stream was started[ the `FrameProducer` waits on a threading event to be set](https://github.com/alliedvision/VimbaPython/blob/master/Examples/multithreading_opencv.py#L165-L174), the [`killswitch` event](https://github.com/alliedvision/VimbaPython/blob/master/Examples/multithreading_opencv.py#L133). This is a threadsafe object that can be set from a different thread which will cause the `FrameProducer.run` method continues and calls [`cam.stop_streaming`](https://github.com/alliedvision/VimbaPython/blob/master/Examples/multithreading_opencv.py#L177).

Untested code that might make my idea clearer and be a decent starting point:
```python
# method to open the context managers, start the streaming, wait for the event, and stop the streaming
def _frame_producer(self):
   with self.vimba:
        with self.cam:
            try:
                self.cam.start_streaming(handler=self.frame_handler, buffer_count=10)
                self._stream_killswitch.wait() # this will wait until the event is set from somewhere
            finally:
                self.cam.stop_streaming()

@Slot()
def start_preview(self):
    self._stream_killswitch = threading.Event() # the event that will be used to stop the streaming
    self._frame_producer_thread = threading.Thread(target=self._frame_producer) # create the thread object to run the frame producer
    self._frame_producer_thread.start() # actually start the thread to execute the method given as target
    self.is_running = True

@Slot()
def stop_preview(self):
    self._stream_killswitch.set() # set the event the producer is waiting on
    self._frame_producer_thread.join() # wait for the thread to actually be done
    self.is_running = False
```

As mentioned, that code was not tested. I hope it is a decent starting point.

**Comment by rk-exxec** [comment]
Thanks for your detailed answer.
I implemented it how you suggested and it seems to be working alright.

Just out of curiosity:
After starting the stream, I force quit the debugging session. 
Then after restarting the program, it says invalid access on the camera. I guess because it was not properly closed.
This is a rare edge case, but do you have a suggestion how to handle or prevent that error?

**Comment by NiklasKroeger-AlliedVision** [comment]
> Then after restarting the program, it says invalid access on the camera. I guess because it was not properly closed.

Yeah that sounds like the camera connection was not closed properly. Do you happen to use Linux? If so I believe the problem might be, that the state of your camera connections is stored in a shared memory section to synchronize it for multiple instances/threads. This is done by using a memory mapped file in `/dev/shm` (filename is some kind of hash that should start with a 3). Once your program is stopped it is save to delete this file. That way the Transport Layer will check to see if a camera is available and not trust some possibly incorrect state in that shared memory. We are currently in the process of preparing a new release where this shared memory is no longer used. So once we release and you install Vimba 4.2 this should no longer happen.

**Comment by rk-exxec** [comment]
>  Do you happen to use Linux?

No, I am using Windows 10 2004. But what you're saying is, that this is not expected behavior?

Pressing start and stop in the vimba viewer elimintates the problem, but the viewer does not show video while doing so.
I am guessing pressing stop in vimba viewer does something that resets the interface to a working state.
Can I emulate that somehow before opening the cam for real?

**Comment by NiklasKroeger-AlliedVision** [comment]
> No, I am using Windows 10

Ah okay. Then the shared memory idea is not relevant for you as no such mechanism is used on Windows.

I did some quick checking and it seems like quitting a debugging session with a force quit does indeed immediately halt your program without calling the `__exit__` functions (checked in VisualStudio Code on Windows 10). So there is no real way to catch that case. One thing you might do instead is stop your code with a KeyboardInterrupt (CTRL-C in the console). This is indeed caught by the context manager and the connection should be closed properly.

Here is a small test script I used to verify this. Perhaps your editor can be configured to send a KeyboardInterrupt instead of halting execution immediately for the force quit functionality? I found [this stackoverflow question](https://stackoverflow.com/questions/45199817/stopping-a-python-process-so-that-context-managers-still-call-exit-in-windo) that at least has an answer for PyCharm... Then you could check that with this example. It should print some kind of exception that was passed to `__exit__` upon exiting. If nothing is printed `__exit__` is not called and there is not really anything that can be done to capture the force quit of the debugger... Perhaps add a more elegant way to stop your program? Since you are using Qt you might be able to just capture a keypress like enter?

```python
import time


class Manager(object):
    def __init__(self, message):
        self.message = message

    def __enter__(self):
        print("Called __enter__")

    def __exit__(self, exc_type, exc_value, exc_traceback):
        print("Called __exit__")
        print(exc_type)
        print(exc_value)
        print(exc_traceback)


if __name__ == "__main__":
    my_manager = Manager("Running")
    with my_manager:
        while True:
            print(my_manager.message)
            time.sleep(1)

```

To be honest I am also a bit confused as to why you are able to open the camera with the vimba viewer to start and stop it. If VimbaPython cannot connect to it, the viewer should not be able to either... They are using the exact same mechanism to talk to the camera...

**Comment by rk-exxec** [comment]
Thanks for the script! I'll see if I can send a KeyboardInterrupt before stopping the debugger.

> If VimbaPython cannot connect to it, the viewer should not be able to either... 

Well, I can connect, but i cannot access the camera features to write them. In case that makes any difference.
```
Traceback (most recent call last):
  File "C:\Users\krr38985\Documents\Python\DropletEval\.venv\lib\site-packages\vimba\feature.py", line 395, in set
    call_vimba_c('VmbFeatureBoolSet', self._handle, self._info.name, as_bool)
  File "C:\Users\krr38985\Documents\Python\DropletEval\.venv\lib\site-packages\vimba\util\tracer.py", line 139, in wrapper
    return func(*args, **kwargs)
  File "C:\Users\krr38985\Documents\Python\DropletEval\.venv\lib\site-packages\vimba\c_binding\vimba_c.py", line 758, in call_vimba_c
    getattr(_lib_instance, func_name)(*args)
  File "C:\Users\krr38985\Documents\Python\DropletEval\.venv\lib\site-packages\vimba\c_binding\vimba_c.py", line 676, in _eval_vmberror
    raise VimbaCError(result)
vimba.c_binding.vimba_common.VimbaCError: VimbaCError(<VmbError.InvalidAccess: -6>)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Users\krr38985\Documents\Python\DropletEval\camera_control.py", line 56, in __init__
    self._setup_camera()
  File "C:\Users\krr38985\Documents\Python\DropletEval\camera_control.py", line 133, in _setup_camera
    self._cam.ReverseY.set(True)
  File "C:\Users\krr38985\Documents\Python\DropletEval\.venv\lib\site-packages\vimba\util\tracer.py", line 139, in wrapper
    return func(*args, **kwargs)
  File "C:\Users\krr38985\Documents\Python\DropletEval\.venv\lib\site-packages\vimba\feature.py", line 412, in set
    raise exc from e
vimba.error.VimbaFeatureError: Invalid access while calling 'set()' of Feature 'ReverseY'. Read access: allowed. Write access: allowed.
```

**Comment by NiklasKroeger-AlliedVision** [comment]
Ah okay, invalid access might mean, that you are simply not allowed to change the setting you are trying to change RIGHT NOW. Probably because when you halted your script while the camera was streaming, and since `__exit__` is not called it was never told to stop streaming. What you could do is to simply execute a `cam.stop_streaming()` (remember you have to be inside the context manager to do this) at the beginning of your camera setup function before trying to change any feature values. If the camera is not streaming there should be no problem (I think). If it was streaming it will stop and you should be able to change your feature values as expected.

I would still recommend trying to find a way to properly close your application and not always rely on the hard stop of the debugger, but this should work around that problem if it occurs.

**Comment by rk-exxec** [comment]
I tried `cam.stop_streaming()` but it didn't unlock the cam.
```
def _setup_camera(self):
    with self._vimba:
        with self._cam:
            self._cam.ExposureTime.set(1000.0)
            self._cam.ReverseY.set(True)
```
Its strange though, that the Exposure setting completes without an error. 
I guess only the ReverseY is not avaiable while running.
I did try the logging feature of the lib and it seems, that `stop_streaming()` does not actually stop the stream because the `camera.is_streaming()` returns false.
Do you know if the vimba viewer has a command logging feature? Would be nice to see how that program can reset the camera.
Also, do you thing running the DeviceReset action command would help?

**Comment by NiklasKroeger-AlliedVision** [comment]
> I tried `cam.stop_streaming()` but it didn't unlock the cam.

Hmm. I will have to try and replicate the issue then I guess. I never asked: **what kind of camera are you using? USB / GigE? Which model?**

> Its strange though, that the Exposure setting completes without an error.
> I guess only the ReverseY is not avaiable while running.

Correct. Exposure is one of the camera settings that can be changed while the camera acquires images. You can see this for example by streaming images with the vimba viewer. While the camera is streaming you are able to change the exposure time, the ReverseY feature (and many others) however cannot be changed. So this might still be something about the streaming/acquisition status of the device...

> Do you know if the vimba viewer has a command logging feature? Would be nice to see how that program can reset the camera.

To be honest I am not sure how you could enable logging in the Vimba viewer. I will see if I can find that out.

> Also, do you thing running the DeviceReset action command would help?

Yes `DeviceReset` should definitely make the camera available again. This is essentially the same as unplugging the camera and plugging it back in. But that also means, that the camera disappears from the USB Bus for a few seconds. So you would have to consider this when you are using that command.

I will try and replicate the problem with the cameras I have available. Maybe I can figure out what is going on that way.

**Comment by NiklasKroeger-AlliedVision** [comment]
Well unfortunately I cannot reproduce the behaviour you are experiencing with the cameras I currently have available. When I halt the test script at the end of this post with the debugger, the camera connection can be opened again and I can change the settings (Exposure Time and ReverseY) as expected. This was tested with a USB Camera (Alvium 1800 U-500m) and a GigE camera (Mako G-507C).

However when I try to reproduce your problem with the Alvium USB camera I do not receive complete frames after halting with the debugger. This is an issue we have experienced before where the camera was unable to finish sending out a frame because communication was suddenly aborted. This leads to a problem where the camera then does not start streaming cleanly until the remaining packages of the previous incomplete frame are "picked up". That is why using the Vimba viewer to start and stop streaming multiple times manages to restart the stream. But from your description this is not the problem you are seeing.

I would honestly suggest you again try to find a way to end your program execution in a "clean" way. Via some kind of button press or even a KeyboardInterrupt. Anything is better than just hard-halting the execution if multiple devices are involved. You are bound to experience some state inconsistencies that are hard to impossible to recover from without a camera device reset. And like I said, a camera device reset is quite disruptive to your program flow as the camera does actually "disappear" from the bus.

Test script I used to halt execution while the camera is streaming:
```python
import threading
import vimba


class Github19(object):
    def __init__(self):
        self._vimba = vimba.Vimba.get_instance()
        with self._vimba:
            self._cam = self._vimba.get_all_cameras()[0]
            print("Is camera streaming? {}".format(self._cam.is_streaming()))

    def setup_camera(self):
        with self._vimba:
            with self._cam:
                self._cam.ExposureTime.set(1000.0)
                self._cam.ReverseY.set(True)

    def start_preview(self):
        self._stream_killswitch = threading.Event()
        self._frame_producer_thread = threading.Thread(target=self._frame_producer)
        self._frame_producer_thread.start()

    def stop_preview(self):
        self._stream_killswitch.set()
        self._frame_producer_thread.join()

    def _frame_producer(self):
        with self._vimba:
            with self._cam:
                try:
                    self._cam.start_streaming(handler=self._frame_handler, buffer_count=10)
                    self._stream_killswitch.wait()
                finally:
                    self._cam.stop_streaming()

    def _frame_handler(self, cam, frame):
        print(frame)
        cam.queue_frame(frame)

if __name__ == "__main__":
    inst = Github19()
    inst.start_preview()
    input("Press <enter> to end cleanly or halt with debugger to end immediately\n")
    inst.stop_preview()
```

**Comment by rk-exxec** [comment]
Thanks again for your answer. I do not plan on halting the debugger during acquisition very often. I just wanted to simulate what happens during a crash or something. And I don't expect the user to know that they should replug the camera when it happens.

I tried something different this time and it seems to work:
```python
def _init_camera(self):
    with self._vimba:
        cams = self._vimba.get_all_cameras()
        self._cam = cams[0]
        with self._cam:
            self._cam.AcquisitionStatusSelector.set('AcquisitionActive')
            if self._cam.AcquisitionStatus.get():
                self._cam.AcquisitionStop.run()
                self._cam.DeviceReset.run()
            else: return
        num_cams = 0
        while num_cams == 0:
            cams = self._vimba.get_all_cameras()
            num_cams = len(cams)
        self._cam = cams[0]
```
The `is_streaming` from the API does not reflect the camera status, only the internal API status. (?)
I think because of that, the `self._cam.stop_streaming()` does not actually send a command to the camera.
So I ask the cam directly and then stop it directly. 
After that, the first acquisition fails at `frame.as_opencv_image()` with `ValueError: 0 is not a valid PixelFormat`.
I guess I am missing some cleanup steps, but I added the reset command and it seems to work.

Thank you very much for your support, you helped me a lot!

**Comment by rk-exxec** [comment]
For anyone interested: 
I fixed the invalid pixel format error without resetting the device which caused some problems itself.
I simply fetched the broken frame after stopping the acquisition. 
This might also execute a proper shutdown of the cam after fetching.
```python
def _init_camera(self):
    with self._vimba:
        cams = self._vimba.get_all_cameras()
        self._cam = cams[0]
        with self._cam:
            self._cam.AcquisitionStatusSelector.set('AcquisitionActive')
            if self._cam.AcquisitionStatus.get():
                self._cam.AcquisitionStop.run()
                # fetch broken frame
                self._cam.get_frame()
```

**Comment by NiklasKroeger-AlliedVision** [comment]
Thank you for providing further information on how you solved your problem!

If you do run into further problems feel free to open a new issue here.
